Ability to filter posts containing keyword in title/body

hjiangsu opened the pull request:
## Pull Request Description

> Review with whitespace off

This PR enables the ability to add and remove keywords. When adding keywords, any post that contains the keyword in the title or body will be filtered out from the feed page (and any widget that uses the feed page).

Some additional notes:
- Added `width: min(MediaQuery.of(context).size.width, 700)` to `showThunderDialog`'s `AlertDialog`. This allows alert dialogs to take up more of the width of the screen similar to the input dialogs (community, user, instance, etc)
- Applied some small UI tweaks to `UserSettingsPage` since it now appears on the Settings page, I've adjusted some of the UI for blocking users/communities/instances to match the other UIs present in the Settings page.


<!--- Please describe what was changed -->

## Issue Being Fixed

<!-- Please describe the problem that is being fixed and, if applicable, reference a GitHub issue -->

Issue Number: #513

## Screenshots / Recordings


https://github.com/thunder-app/thunder/assets/30667958/639c179d-8b96-4a50-95e2-847ee87bf446

<!-- This section is optional but highly recommended to show off your changes! -->

## Checklist

- [x] Did you update CHANGELOG.md?
- [x] Did you use localized strings where applicable?
- [x] Did you add `semanticLabel`s where applicable for accessibility?

hjiangsu commented:
@micahmo ready for review!

micahmo commented on lib/user/pages/user_settings_page.dart:
```diff
@@ -163,25 +163,28 @@ class _UserSettingsPageState extends State<UserSettingsPage> {
-                          ),
+                      Padding(
+                        padding: const EdgeInsets.symmetric(horizontal: 16.0, vertical: 8.0),
+                        child: Row(
```
The only tiny problem with these changes is that the plus button (to add a new block) and the x button (to delete a block) are no longer aligned. It's very minor though. 😊

hjiangsu replied:
I think this should be okay for now! I do think the plus button now lines up with the ToggleOptions so maybe its just a matter of adjusting the x button to be closer to the right

micahmo replied:
> maybe its just a matter of adjusting the x button to be closer to the right

Agreed!

micahmo reviewed (approved):
Awesome feature!

One thing I noticed while testing... This might be a problem I introduced with my dialog refactor, but I was able to add an empty string as a filter by pressing enter on the keyboard. (I think this is a problem with the comment search too.)

hjiangsu commented:
> I  was able to add an empty string as a filter by pressing enter on the keyboard

Ahh, I missed that edge case while testing it. I can add in an additional check to make sure we can't add in empty strings (and also trim whitespaces while I'm at it)

hjiangsu commented on lib/feed/utils/post.dart:
```diff
@@ -39,6 +46,16 @@ Future<Map<String, dynamic>> fetchPosts({
+        final title = postView.post.name.toLowerCase();
+        final body = postView.post.body?.toLowerCase() ?? '';
+
+        return !keywordFilters.any((keyword) => title.contains(keyword.toLowerCase()) || body.contains(keyword.toLowerCase()));
```
Yeah, I'm hoping this isn't slow either 😅 This is the side-effect of Lemmy not having an internal implementation for keyword filtering so I hope they add it in sometime in the future to reduce the need for this!

hjiangsu commented on lib/settings/pages/filter_settings_page.dart:
```diff
@@ -0,0 +1,149 @@
+                alignment: Alignment.centerLeft,
+                child: keywordFilters.isEmpty
+                    ? Text(
+                        l10n.noKeywordFilters,
```
I was trying to match it with the other placeholder text we have, but I understand what you mean here 😅

hjiangsu replied:
I moved the description for the filter to the top of the page. Hopefully this does the trick! b3aa2eb11f39559797c9815899af33c1915aee47

micahmo replied:
Yeah this looks better!

hjiangsu commented on lib/settings/pages/filter_settings_page.dart:
```diff
@@ -0,0 +1,149 @@
+                                  Navigator.of(context).pop();
+                                },
+                                secondaryButtonText: l10n.cancel,
+                                onSecondaryButtonPressed: (context) => Navigator.of(context).pop(),
```
Done! 8dc3efd9da72921a34281bbc102a4c79a0c5f440

hjiangsu commented:
Added trimming to `showInputDialog` to disable the button when there's empty strings, and added trimming to keywords in eff689a5f0a4cc4ddd29df89def6389125dedd46!

micahmo commented:
> Added trimming to `showInputDialog` to disable the button when there's empty strings, and added trimming to keywords in [eff689a](https://github.com/thunder-app/thunder/commit/eff689a5f0a4cc4ddd29df89def6389125dedd46)!

Nice! We can try to add something similar to comment search at some point.
